### PR TITLE
[PATCH v5] api: sync: new memory barriers

### PR DESCRIPTION
**Pull request by psavol**
Added data synchronizing memory barriers that are stronger than the current barriers, and can be used also when accessing memory mapped device registers.

v4: rebased

### CODE REVIEW

File: include/odp/api/spec/sync.h [hunk 1]
```diff
@@ -23,15 +23,24 @@ extern "C" {
...
- * synchronization mechanisms (e.g. execution barriers, locks, queues, etc )
- * include all necessary memory barriers, so these calls are not needed when
- * using those. Also ODP atomic operations have memory ordered versions. These
- * explicit barriers may be needed when thread synchronization is based on
- * a non-ODP defined mechanism. Depending on the HW platform, heavy usage of
- * memory barriers may cause significant performance degradation.
+ * A memory barrier enforces the order between memory accesses (load and/or store instructions)
+ * specified (in program order) before the barrier with those specified after the barrier.
+ * A barrier may affect both compiler optimizations and CPU out-of-order execution. Depending
+ * on the HW platform and the barrier used, heavy usage of barriers may cause significant
```
[the remark this answers is not in the document]
Comment (psavol) [reply include/odp/api/spec/sync.h:29]: Fixed in v2. Slightly different wording.

File: include/odp/api/spec/sync.h [hunk 2]
```diff
@@ -23,15 +23,24 @@ extern "C" {
...
+ * performance degradation.
+ *
+ * Application needs to use these memory barrier functions e.g. when threads synchronize
+ * directly through shared memory (without ODP defined mechanisms) or when a thread accesses
+ * memory mapped registers of a device.
+ *
+ * Application threads do not need to use explicit memory barriers when using ODP synchronization
+ * mechanisms (execution barriers, locks, etc), or when exchanging data through other ODP
+ * mechanisms (queue, stash, etc). Those ODP calls include necessary (acquire and release) memory
```
[the remark this answers is not in the document]
Comment (psavol) [reply include/odp/api/spec/sync.h:38]: Fixed in v2

File: include/odp/api/spec/sync.h [hunk 3]
```diff
@@ -79,6 +88,43 @@ void odp_mb_acquire(void);
  */
 void odp_mb_full(void);
 
+/**
+ * Memory barrier for load and store synchronization
+ *
+ * This memory barrier ensures that all memory accesses (load and store instructions)
+ * specified before the barrier (in program order) are complete prior any memory access
```
[the remark this answers is not in the document]
Comment (psavol) [reply include/odp/api/spec/sync.h:95]: Fixed in v2

File: platform/linux-generic/include/odp/api/plat/sync_inlines.h
Comment (MatiasElo) [inline platform/linux-generic/include/odp/api/plat/sync_inlines.h:1]: Nokia's copyright can be added.

File: include/odp/api/spec/sync.h [hunk 4]
```diff
@@ -23,15 +23,24 @@ extern "C" {
...
- * a non-ODP defined mechanism. Depending on the HW platform, heavy usage of
- * memory barriers may cause significant performance degradation.
+ * A memory barrier enforces the order between memory accesses (load and/or store instructions)
+ * specified (in program order) before the barrier with those specified after the barrier.
+ * A barrier may affect both compiler optimizations and CPU out-of-order execution. Depending
+ * on the HW platform and the barrier used, heavy usage of barriers may cause significant
+ * performance degradation.
+ *
+ * Application needs to use these memory barrier functions e.g. when threads synchronize
+ * directly through shared memory (without ODP defined mechanisms) or when a thread accesses
```
Comment (JereLeppanen) [inline include/odp/api/spec/sync.h:32]: The application may or may not "need" this, and also "when threads synchronize" doesn't just happen, it's something that the application is actively doing. So maybe something like "Application may use these ... functions e.g. to synchronize threads directly ... or to access memory mapped registers of a device."

File: include/odp/api/spec/sync.h [hunk 5]
```diff
@@ -79,6 +88,43 @@ void odp_mb_acquire(void);
...
+ * also memory access completion is ensured. The barrier may be useful e.g. when synchronizing
+ * loads and stores into memory mapped registers of a device.
+ */
+void odp_mb_sync(void);
+
+/**
+ * Memory barrier for load synchronization
+ *
+ * This memory barrier ensures that all load instructions specified before the barrier
+ * (in program order) complete prior to any load instructions specified after
```
[the remark this answers is not in the document]
Comment (psavol) [reply include/odp/api/spec/sync.h:108]: Fixed in v2. The intention is that all load/store instructions are complete, before any single load/store instruction begins execution.

**Review by JereLeppanen — APPROVED** [review]
I only have couple of nitpicks. I leave it up to you whether you want to re-word anything or not - approved regardless.

File: platform/linux-generic/arch/aarch64/odp/api/abi/sync_inlines.h [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright (c) 2023 Nokia
+ */
+
+#ifndef ODP_ARCH_SYNC_INLINES_H_
+#define ODP_ARCH_SYNC_INLINES_H_
```
Comment (MatiasElo) [inline platform/linux-generic/arch/aarch64/odp/api/abi/sync_inlines.h:6]: Looks like a slightly more common naming scheme would be to use `ODP_API_ABI_SYNC_INLINES_H_`, which better matches the file path, but no strong feelings about this.

File: include/odp/api/spec/sync.h [hunk 7]
```diff
@@ -23,15 +23,24 @@ extern "C" {
...
- * specified before and after the barrier. These barriers may affect both
- * compiler optimizations and CPU out-of-order execution. All ODP
- * synchronization mechanisms (e.g. execution barriers, locks, queues, etc )
- * include all necessary memory barriers, so these calls are not needed when
- * using those. Also ODP atomic operations have memory ordered versions. These
- * explicit barriers may be needed when thread synchronization is based on
- * a non-ODP defined mechanism. Depending on the HW platform, heavy usage of
- * memory barriers may cause significant performance degradation.
+ * A memory barrier enforces the order between memory accesses (load and/or store instructions)
+ * specified (in program order) before the barrier with those specified after the barrier.
```
[the remark this answers is not in the document]
Comment (psavol) [reply include/odp/api/spec/sync.h:26]: Agree, removed "instruction" in v3

File: platform/linux-generic/include/odp/api/plat/sync_inlines.h
Comment (psavol) [reply platform/linux-generic/include/odp/api/plat/sync_inlines.h:1]: Fixed in v3

File: include/odp/api/spec/sync.h
[the remark this answers is not in the document]
Comment (psavol) [reply include/odp/api/spec/sync.h:1]: Fixed in v3

File: platform/linux-generic/arch/aarch64/odp/api/abi/sync_inlines.h (hunk 6 above)
Comment (psavol) [reply platform/linux-generic/arch/aarch64/odp/api/abi/sync_inlines.h:6]: Keeping ARCH as it's more readable and there's another file with that in the same directory already.

File: include/odp/api/spec/sync.h (hunk 4 above)
Comment (psavol) [reply include/odp/api/spec/sync.h:32]: Fixed in v3. Section rewritten along those lines.

**Pull request merged**